### PR TITLE
[apache/helix] -- Adjusted the missingTopStatePartitionsBeyondThresholdGuage when top-state Instance recovers

### PR DESCRIPTION
**Pull request by himanshukandwal**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:
#2590

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

We have introduced an additional step in the `TopStateComesBack` code-path, where we are decrementing the `missingTopStatePartitionsBeyondThresholdGuage` when the associated `MissingTopStateRecord` has been previously marked as failed. The marking of failure on record can only happen of we have not seen top-state coming back before threshold and the `missingTopStatePartitionsBeyondThresholdGuage` was previously incremented. 

### Tests

- [X] The following tests are written for this issue:
* Updated TestTopStateHandoffMetrics.testTopStateFailedHandoff
* Added TestTopStateHandoffMetrics.testTopStateFailedUnrecoveredHandoff

- The following is the result of the "mvn test" command on the appropriate module:
(If CI test fails due to known issue, please specify the issue and test PR locally. Then copy & paste the result of "mvn test" to here.)
```java
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 12, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- jacoco:0.8.6:report (generate-code-coverage-report) @ helix-core ---
[INFO] Loading execution data file /Users/hkandwal/Documents/workspaces/projects/helix_os_hk/helix-core/target/jacoco.exec
[INFO] Analyzed bundle 'Apache Helix :: Core' with 799 classes
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  25.029 s
[INFO] Finished at: 2023-08-11T16:02:07-07:00
[INFO] ------------------------------------------------------------------------
```

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/controller/stages/TopStateHandoffReportStage.java [hunk 1]
```diff
@@ -491,6 +491,12 @@ private void reportTopStateComesBack(ResourceControllerDataProvider cache, Map<S
                 true);
       }
     }
+
+    // In case of recovery after failure, we should decrement the missingTopStateBeyondThresholdGauge value.
+    if (clusterStatusMonitor != null && record.isFailed()) {
```
Comment (desaikomal) [inline helix-core/src/main/java/org/apache/helix/controller/stages/TopStateHandoffReportStage.java:496]: First and foremost, congratulations on your first change.
Q: why do we decrement only for recovery after failure scenario and not in-general recovery?
Q: why don't i see change for 1min related reset window?

**Review by rahulrane50 — APPROVED** [review]
LGTM1 Congrats on your first change! Great work!

File: helix-core/src/main/java/org/apache/helix/controller/stages/TopStateHandoffReportStage.java (hunk 1 above)
Comment (desaikomal) [reply helix-core/src/main/java/org/apache/helix/controller/stages/TopStateHandoffReportStage.java:496]: Had offline conversation. A: to first one is that whenever we cross threshold, we mark it FAILED.
A: to the second one is:this is generic reset - not specific to the metric in question.

**Review by desaikomal — APPROVED** [review]
thanks

**Comment by himanshukandwal** [comment]
This PR has been reviewed and approved by @desaikomal and @rahulrane50 and is ready to be merged

Final Commit Message:
Adjusted decrementing the missingTopStatePartitionsBeyondThresholdGuage when top-state Instance recovers after the threshold failure.

File: helix-core/src/main/java/org/apache/helix/controller/stages/TopStateHandoffReportStage.java (hunk 1 above)
Comment (rahulrane50) [reply helix-core/src/main/java/org/apache/helix/controller/stages/TopStateHandoffReportStage.java:496]: Also to second one, we want to keep fallback backup reset mechanism to reset the metric.

**Review by xyuanlu — APPROVED** [review]

**Pull request merged**